TextView extensions

jaredsburrows opened the pull request:
Help add extensions and tests for https://github.com/android/android-ktx/issues/227 @JakeWharton @romainguy

dovahkiin98 commented:
all these extensions are already present in PR #269

illuzor commented:
I think **content()** must be val with getter/setter, and **isEmpty()** must be val with getter.
Also I suggest to add **isBlank** property/

JakeWharton commented on src/main/java/androidx/widget/TextView.kt:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Returns `true` if this char sequence is empty (contains no characters).
+ */
+fun TextView.isEmpty() = text.isEmpty()
```
I don't find these two or `clear()` that useful.

jaredsburrows replied:
I will remove.

jaredsburrows replied:
Fixed.

JakeWharton replied:
Forget to push up changes?

jaredsburrows replied:
I removed both `isEmpty()` and `clear()`.

JakeWharton commented on src/main/java/androidx/widget/TextView.kt:
```diff
@@ -0,0 +1,57 @@
+ * Adds a [TextWatcher] to the list of those whose methods are called
+ * whenever this [TextView]'s text changes.
+ */
+fun TextView.textChange(
```
This should be `inline` and its lambdas `crossinline` and it should return the `TextWatcher` instance to allow unregistering.

jaredsburrows replied:
Fixed.

JakeWharton commented on src/main/java/androidx/widget/TextView.kt:
```diff
@@ -0,0 +1,57 @@
+        override fun afterTextChanged(s: Editable) = after.invoke(s)
+
+        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) =
+            before.invoke(s.toString(), start, count, after)
```
Remove `toString()`

jaredsburrows replied:
Ok.

jaredsburrows replied:
Fixed.

JakeWharton commented on src/main/java/androidx/widget/TextView.kt:
```diff
@@ -0,0 +1,57 @@
+            before.invoke(s.toString(), start, count, after)
+
+        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) =
+            onTextChanged(s.toString(), start, before, count)
```
Remove `toString()`

jaredsburrows replied:
Ok.

jaredsburrows replied:
Fixed.

jaredsburrows commented:
@JakeWharton all fixed!

jaredsburrows commented:
Anything else you want me to fix in this PR @JakeWharton? :)

dovahkiin98 commented:
@jaredsburrows Why did you remove isEmpty and clear()? 
if you think they're "too obvious", ViewGroup has isEmpty and size which are too obvious too

jaredsburrows commented:
@dovahkiin98 @JakeWharton asked me to.

jaredsburrows commented:
Can we merge? @JakeWharton? :)

JakeWharton commented:
Every time you ask you get dropped to the bottom of the queue. We are aware
of the open issues and PRs and don't need reminding.

On Mon, Mar 5, 2018 at 3:05 PM Jared Burrows <notifications@github.com>
wrote:

> Can we merge? @JakeWharton <https://github.com/jakewharton>? :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/android/android-ktx/pull/386#issuecomment-370546433>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEWsqpPkbOWVqsXh3XPpKkf1Fsdr6ks5tbZprgaJpZM4SZeYL>
> .
>

romainguy commented on src/main/java/androidx/widget/TextView.kt:
```diff
@@ -0,0 +1,55 @@
+ * Adds a [TextWatcher] to the list of those whose methods are called
+ * whenever this [TextView]'s text changes.
+ */
+inline fun TextView.textChange(
```
We need a better name here. Maybe addTextListener() and then add doOnBeforeTextChanged/doOnAfterTextChanged() like in the Animator.kt file?

jaredsburrows replied:
I agree with the name change. Let me make the update.

jaredsburrows replied:
I renamed `testChange()` to `addTextListener()` and added both `doOnBeforeTextChanged()` and `doOnAfterTextChanged()` with tests!

jaredsburrows commented:
@romainguy @JakeWharton All updated!

Tougee commented:
Didn't see your updates, I update base on your commit @jaredsburrows [here](https://github.com/android/android-ktx/pull/433)

jaredsburrows commented:
@romainguy I updated the build but I am not sure how this error affected my branch:
```
#!/bin/bash -eo pipefail
./gradlew check --no-daemon
:checkApi


No source files specified: recursively including all sources found in the source path
...
api/current.txt:654: error: Removed class androidx.core.widget.TextViewKt [RemovedClass:8]

warning: Added package androidx.widget [AddedPackage:2]Aborting: Found compatibility problems with --check-compatibility
:checkApi FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':checkApi'.
> Process 'command '/usr/lib/jvm/java-8-openjdk-amd64/bin/java'' finished with non-zero exit value 255

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 22s
1 actionable task: 1 executed
Exited with code 1
```

jaredsburrows commented:
@romainguy @JakeWharton How can I get the CI `test_instrumented` to pass?

JakeWharton commented:
You cannot. #141

On Thu, Mar 22, 2018 at 3:29 AM Jared Burrows <notifications@github.com>
wrote:

> @romainguy <https://github.com/romainguy> @JakeWharton
> <https://github.com/jakewharton> How can I get the CI test_instrumented
> to pass?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/android/android-ktx/pull/386#issuecomment-375203860>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEalIfW-xS3WZGBoJoTmMFT-3ZfRIks5tg1LDgaJpZM4SZeYL>
> .
>

jaredsburrows commented:
@romainguy @JakeWharton Review :)

JakeWharton commented:
Text watcher-based listeners have been merged in http://r.android.com/720364
